test/client_reset_test.dart

```diff
@@ -0,0 +1,256 @@
+
+import 'package:test/test.dart' hide test, throws;
+import '../lib/realm.dart';
+import '../lib/src/configuration.dart' show ClientResetHandlerInternal, ClientResyncModeInternal, BeforeResetCallback, AfterResetCallback, ClientResetCallback;
+import 'test.dart';
+
+Future<void> main([List<String>? args]) async {
+  await setupTests(args);
+
+  baasTest("Configuration.flexibleSync set recoverOrDiscard as a default resync mode", (appConfiguration) async {
+    final app = App(appConfiguration);
+    final user = await getIntegrationUser(app);
+    expect(
+        Configuration.flexibleSync(
+          user,
+          [Task.schema, Schedule.schema],
+          clientResetHandler: ManualRecoveryHandler((syncError) {}),
+        ).clientResetHandler.clientResyncMode,
+        ClientResyncModeInternal.manual);
+    expect(
+        Configuration.flexibleSync(
+          user,
+          [Task.schema, Schedule.schema],
+          clientResetHandler: const DiscardUnsyncedChangesHandler(),
+        ).clientResetHandler.clientResyncMode,
+        ClientResyncModeInternal.discardLocal);
+    expect(
+        Configuration.flexibleSync(
+          user,
+          [Task.schema, Schedule.schema],
+          clientResetHandler: const RecoverUnsyncedChangesHandler(),
+        ).clientResetHandler.clientResyncMode,
+        ClientResyncModeInternal.recover);
+
+    expect(
+        Configuration.flexibleSync(
+          user,
+          [Task.schema, Schedule.schema],
+          clientResetHandler: const RecoverOrDiscardUnsyncedChangesHandler(),
+        ).clientResetHandler.clientResyncMode,
+        ClientResyncModeInternal.recoverOrDiscard);
+
+    expect(Configuration.flexibleSync(user, [Task.schema, Schedule.schema]).clientResetHandler.clientResyncMode, ClientResyncModeInternal.recoverOrDiscard);
+  });
+
+  baasTest('ManualRecoveryHandler error is reported in callback', (appConfig) async {
+    final app = App(appConfig);
+    final user = await getIntegrationUser(app);
+
+    final resetCompleter = Completer<ClientResetError>();
+    final config = Configuration.flexibleSync(
+      user,
+      [Task.schema, Schedule.schema],
+      clientResetHandler: ManualRecoveryHandler((syncError) {
+        resetCompleter.complete(syncError);
+      }),
+    );
+
+    final realm = await Realm.open(config);
+    await realm.syncSession.waitForUpload();
+
+    await triggerClientReset(realm);
+
+    final error = await resetCompleter.future;
+    expect(error.message, contains('Bad client file identifier'));
+  });
+
+  for (Type clientResetHandlerType in [
+    RecoverOrDiscardUnsyncedChangesHandler,
+    RecoverUnsyncedChangesHandler,
+    DiscardUnsyncedChangesHandler,
+  ]) {
+    baasTest('$clientResetHandlerType.manualResetFallback invoked when throw an error on Before Callback', (appConfig) async {
+      final app = App(appConfig);
+      final user = await getIntegrationUser(app);
+
+      final onManualResetFallback = Completer<void>();
```

Comment:
Can we change this to `Completer<ClientResetError>` and then assert on the error we get from the manual reset fallback?